FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/AzureTableMetadataStore.cs

```diff
@@ -0,0 +1,169 @@
+        }
+
+        public async Task<Dictionary<string, long>> GetPatientVersionsAsync(byte queueType, List<string> patientsHash, CancellationToken cancellationToken = default)
+        {
+            var pk = TableKeyProvider.CompartmentPartitionKey(queueType);
+            var patientVersions = new Dictionary<string, long>();
+
+            for (var i = 0; i < patientsHash.Count(); i += MaxCountOfQueryEntities)
+            {
+                var selectedPatients = patientsHash.Skip(i).Take(MaxCountOfQueryEntities).ToList();
+                var jobEntityQueryResult = _metadataTableClient.QueryAsync<CompartmentInfoEntity>(
+                    filter: TransactionGetByKeys(pk, selectedPatients),
+                    cancellationToken: cancellationToken);
+
+                await foreach (var pageResult in jobEntityQueryResult.AsPages().WithCancellation(cancellationToken))
+                {
+                    foreach (var entity in pageResult.Values)
+                    {
+                        patientVersions[entity.RowKey] = entity.VersionId;
+                    }
+                }
+            }
+
+            return patientVersions;
+        }
+
+        public async Task<Dictionary<string, long>> GetPatientVersionsAsync(byte queueType, CancellationToken cancellationToken = default)
+        {
+            var patientVersions = new Dictionary<string, long>();
+
+            var jobEntityQueryResult = _metadataTableClient.QueryAsync<CompartmentInfoEntity>(
+                filter: $"PartitionKey eq '{TableKeyProvider.CompartmentPartitionKey(queueType)}'",
+                cancellationToken: cancellationToken);
+
+            await foreach (var pageResult in jobEntityQueryResult.AsPages().WithCancellation(cancellationToken))
+            {
+                foreach (var entity in pageResult.Values)
+                {
+                    patientVersions[entity.RowKey] = entity.VersionId;
+                }
+            }
+
+            return patientVersions;
+        }
+
+        public async Task UpdatePatientVersionsAsync(byte queueType, Dictionary<string, long> patientVersions, CancellationToken cancellationToken = default)
+        {
+            var transactionActions = patientVersions
+                .Select(patientVersion => new TableTransactionAction(TableTransactionActionType.UpsertReplace, new CompartmentInfoEntity
+                {
+                    PartitionKey = TableKeyProvider.CompartmentPartitionKey(queueType),
+                    RowKey = patientVersion.Key,
+                    VersionId = patientVersion.Value,
+                })).ToList();
+
+            for (var i = 0; i < patientVersions.Count; i += MaxCountOfTransactionEntities)
+            {
+                var selectedTransactionActions = transactionActions.Skip(i).Take(MaxCountOfTransactionEntities).ToList();
```

Comment:
It will get all the data no matter the `patientVersions` can be divided by `MaxCountOfTransactionEntities` or not, [Reference](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.take?view=net-6.0)